README.md
py-videodownloader
===================
Python module and script for downloading video source files from the major
online streaming sites (YouTube, Vimeo, etc.)


Usage
=====
Videos from various providers can be specified using the following command line
format:

    Usage: videodownloader.py -p PROVIDER [-f FMT] [-d DIR] videoID [... videoID]

    Options:
      -h, --help            show this help message and exit
      -e EXT, --ext=EXT     Manually override video extension.
      -f FORMAT, --format=FORMAT
                            Format of video to download. Run with no video IDs for
                            a provider specific list.
      -t TITLE, --title=TITLE
                            Manually override video title.
      -p PROVIDER, --provider=PROVIDER
                            Online provider from where to download the video.
                            (Available: 'Vimeo', 'YouTube')
      --debug               Enable debugging output.


Example
-------
The following two commands will download a 720p video from Vimeo and a 1080p
one from YouTube.

    $ videodownloader.py -p Vimeo -f hd 5720832
    $ videodownloader.py -p YouTube -f 37 tgbNymZ7vqY

To see a list of the formats supported by a provider run the command without
any video IDs.

    $ videodownloader.py -p YouTube
    videodownloader-2.0.0 - by Jake Wharton <jakewharton@gmail.com>

    Format     Description
    ---------- ----------------------------------------
    13         176x144 3GP/AMR Mono 3GP
    17         176x144 3GP/AAC Mono 3GP
    22         1280x720 H.264/AAC Stereo MP4
    18         480x360/480x270 H.264/AAC Stereo MP4
    37         1920x1080 H.264/AAC Stereo MP4
    35         640x480/640x360 H.264/AAC Stereo FLV
    34         320x240 H.264/AAC Stereo FLV
    5          320x240 H.263/MP3 Mono FLV
    6          320x240 H.263/MP3 Mono FLV

Omitting a format will automatically download the best format available for
that video from the provider.


Developed By
============
* Jake Wharton - <jakewharton@gmail.com>

Git repository located at
[http://github.com/JakeWharton/py-videodownloader](http://github.com/JakeWharton/py-videodownloader)


Thanks
======
*   __chexov__

    For `py-youtube-downloader`, the inspiration and basis for this module and
    script. Check it out at
    [http://github.com/chexov/py-youtube-downloader](http://github.com/chexov/py-youtube-downloader).

*   __Arlo Faria__

    Improvement of the token, resolution, and title parsing for YouTube videos.


License
=======
    Copyright 2010 Jake Wharton

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.


CHANGELOG.md
Change Log
==========

Version 2.0.0 *(2010-05-19)*
--------------------------------
Rewrite of a significant portion of the codebase to be more modular and easier to extend.

New Features:

 * Supports thumbnail URL retrieval.
 * Improved Vimeo information parsing.
 * All data is stored in instance properties.
 * Provider-specific metadata is now retrieved.
 * Specify custom title and extension for videos.

Bug Fixes:

 * Tests properly reference the library after the folder shuffle.

Version 1.2.4 *(2010-05-07)*
----------------------------
Bug Fixes:

 * Convert all files to Unix-style line endings.

Version 1.2.3 *(2010-05-04)*
----------------------------
New Features:

 * Installable via `easy_install` or `pip`!

Version 1.2.2 *(2010-04-23)*
----------------------------
New Features:

 * New method for parsing token, resolutions, and the title for YouTube videos.

Version 1.2.1 *(2010-04-23)*
----------------------------
Bug Fixes:

 * Add period between file name and extension for a proper full filename.

Version 1.2.0 *(2010-04-21)*
----------------------------
New Features:

 * Ability to determine valid formats for a specific video.
 * Automatic downloading of highest format available if none specified.

Version 1.1.0 *(2010-04-04)*
--------------------------------
New Features:

 * Dynamic loading of providers.
 * Support passing in formats.
 * Unit test for all providers.
 * Alternate destination directory support.

Version 1.0.0 *(2010-03-05)*
----------------------------
Initial release with Vimeo and YouTube support.


videodownloader/providers/youtube.py
__license__ = '''
Copyright 2010 Jake Wharton

py-video-downloader is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as
published by the Free Software Foundation, either version 3 of
the License, or (at your option) any later version.

py-video-downloader is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General
Public License along with py-video-downloader.  If not, see
<http://www.gnu.org/licenses/>.
'''

from videodownloader.providers import Provider
import re
import urllib

class YouTube(Provider):
    FORMAT_PRIORITY = ['37', '22', '35', '18', '34', '5']
    FORMATS = {
        '5' : '320x240 H.263/MP3 Mono FLV',
        '13': '176x144 3GP/AMR Mono 3GP',
        '17': '176x144 3GP/AAC Mono 3GP',
        '18': '480x360/480x270 H.264/AAC Stereo MP4',
        '22': '1280x720 H.264/AAC Stereo MP4',
        '34': '320x240 H.264/AAC Stereo FLV',
        '35': '640x480/640x360 H.264/AAC Stereo FLV',
        '37': '1920x1080 H.264/AAC Stereo MP4',
    }

    def __init__(self, id, **kwargs):
        super(YouTube, self).__init__(id, **kwargs)

        #Load video meta information
        url  = 'http://youtube.com/get_video_info?video_id=%s' % self.id
        self._debug('YouTube', '__init__', 'Downloading "%s"...' % url)
        self._html = super(YouTube, YouTube)._download(url).read()

        #Get available formats
        self.formats = set()
        for match in re.finditer(r'itag%3D(\d+)', self._html):
            if match.group(1) not in YouTube.FORMATS.keys():
                print 'WARNING: Unknown format "%s" found.'
            self.formats.add(match.group(1))
        self._debug('YouTube', '__init__', 'formats', ', '.join(self.formats))

        #Get video title if not explicitly set
        if self.title is id:
            match = re.search(r'&title=(.+?)$', self._html, re.DOTALL)
            if match:
                self.title = urllib.unquote_plus(match.group(1))
                self._debug('YouTube', '__init__', 'title', self.title)
        self._debug('YouTube', '__init__', 'title', self.title)

        #Get video filename if not explicity set
        self.filename = self.title if self.filename is None else self.filename
        self._debug('YouTube', '__init__', 'filename', self.filename)

        #Get proper file extension if a valid format was supplied
        if self.format is not None and self.format in YouTube.FORMATS.keys():
            self.fileext = YouTube.FORMATS[self.format][-3:].lower()
            self._debug('YouTube', '__init__', 'fileext', self.fileext)

        #Get magic data needed to download
        match = re.search(r'&token=([-_0-9a-zA-Z]+%3D)', self._html)
        self.token = urllib.unquote(match.group(1)) if match else None
        self._debug('YouTube', '__init__', 'token', self.token)

        #Video thumbnail
        match = re.search(r'&thumbnail_url=(.+?)&', self._html)
        self.thumbnail = urllib.unquote(match.group(1)) if match else None
        self._debug('YouTube', '__init__', 'thumbnail', self.thumbnail)

        #Video duration (seconds)
        try:
            match = re.search(r'&length_seconds=(\d+)&', self._html)
            self.duration = int(match.group(1)) if match else -1
        except ValueError:
            #TODO: warn
            self.duration = -1
        self._debug('YouTube', '__init__', 'duration', self.duration)

        #Other YouTube-specific information
        match = re.search(r'&author=(.+?)&', self._html)
        self.author = match.group(1) if match else None
        self._debug('YouTube', '__init__', 'author', self.author)

        match = re.search(r'keywords=(.+?)&', self._html)
        self.keywords = set(urllib.unquote(match.group(1)).split(',')) if match else set([])
        self._debug('YouTube', '__init__', 'keywords', ','.join(self.keywords))

        try:
            match = re.search(r'&avg_rating=(\d\.\d+)&', self._html)
            self.rating = float(match.group(1)) if match else -1.0
        except ValueError:
            #TODO: warn
            self.rating = -1.0
        self._debug('YouTube', '__init__', 'rating', self.rating)


    def get_download_url(self):
        #Validate format
        if self.format is None:
            self.format = self._get_best_format()
        elif self.format not in self.formats:
            raise ValueError('Invalid format "%s". Valid formats are "%s".' % (self.format, '", "'.join(self.formats)))

        #Check extension
        if self.fileext is None or self.fileext == Provider.DEFAULT_EXT:
            self.fileext = YouTube.FORMATS[self.format][-3:].lower()

        url = 'http://youtube.com/get_video?video_id=%s&fmt=%s&t=%s' % (self.id, self.format, self.token)

        self._debug('YouTube', 'get_download_url', 'url', url)
        return url

    def _get_best_format(self):
        for format in YouTube.FORMAT_PRIORITY:
            if format in self.formats:
                self._debug('YouTube', '_get_best_format', 'format', format)
                return format
        raise ValueError('Could not determine the best available format. YouTube has likely changed its page layout. Please contact the author of this script.')